From 600f659c2d0453e98eca413152776308e1b8c8cd Mon Sep 17 00:00:00 2001
From: "Petar G. Georgiev" <petarg@codeaurora.org>
Date: Fri, 29 Nov 2019 14:56:55 +0200
Subject: [CHANGE 5/5] waylandsink: Forcibly release pending buffers on PAUSED
 state

Add functionality to forcibly release any pending buffers in
the wayland compositor when changing the state from PLAYING to
PAUSED. This is needed because some live sources may require
all buffers to be returned when transitioning between those 2
states.
---
 ext/wayland/gstwaylandsink.c |  5 ++++-
 ext/wayland/wldisplay.c      | 14 ++++++++++++++
 ext/wayland/wldisplay.h      |  1 +
 3 files changed, 19 insertions(+), 1 deletion(-)

diff --git a/ext/wayland/gstwaylandsink.c b/ext/wayland/gstwaylandsink.c
index 9bf51b3..6fc7c0b 100644
--- a/ext/wayland/gstwaylandsink.c
+++ b/ext/wayland/gstwaylandsink.c
@@ -460,8 +460,11 @@ gst_wayland_sink_change_state (GstElement * element, GstStateChange transition)
     return ret;
 
   switch (transition) {
-    case GST_STATE_CHANGE_PAUSED_TO_READY:
+    case GST_STATE_CHANGE_PLAYING_TO_PAUSED:
+      gst_wl_display_force_release_buffers (sink->display);
       gst_buffer_replace (&sink->last_buffer, NULL);
+      break;
+    case GST_STATE_CHANGE_PAUSED_TO_READY:
       if (sink->window) {
         if (gst_wl_window_is_toplevel (sink->window)) {
           g_clear_object (&sink->window);
diff --git a/ext/wayland/wldisplay.c b/ext/wayland/wldisplay.c
index 5a9fece..2a94dfb 100644
--- a/ext/wayland/wldisplay.c
+++ b/ext/wayland/wldisplay.c
@@ -456,3 +456,17 @@ gst_wl_display_unregister_buffer (GstWlDisplay * self, gpointer buf)
     g_hash_table_remove (self->buffers, buf);
   g_mutex_unlock (&self->buffers_mutex);
 }
+
+void
+gst_wl_display_force_release_buffers (GstWlDisplay * self)
+{
+  /* to avoid buffers being unregistered from another thread
+   * at the same time, take their ownership */
+  g_mutex_lock (&self->buffers_mutex);
+  g_hash_table_foreach (self->buffers, (GHFunc) g_object_ref, NULL);
+  g_mutex_unlock (&self->buffers_mutex);
+
+  g_hash_table_foreach (self->buffers,
+      (GHFunc) gst_wl_buffer_force_release_and_unref, NULL);
+  g_hash_table_remove_all (self->buffers);
+}
diff --git a/ext/wayland/wldisplay.h b/ext/wayland/wldisplay.h
index accc96a..35bb5f5 100644
--- a/ext/wayland/wldisplay.h
+++ b/ext/wayland/wldisplay.h
@@ -94,6 +94,7 @@ GstWlDisplay *gst_wl_display_new_existing (struct wl_display * display,
 /* see wlbuffer.c for explanation */
 void gst_wl_display_register_buffer (GstWlDisplay * self, gpointer buf);
 void gst_wl_display_unregister_buffer (GstWlDisplay * self, gpointer buf);
+void gst_wl_display_force_release_buffers (GstWlDisplay * self);
 
 gboolean gst_wl_display_check_format_for_shm (GstWlDisplay * display,
     GstVideoFormat format);
-- 
1.9.1

